src/Mir/FancyMuxTree.hs

```diff
@@ -0,0 +1,428 @@
+import           Data.Text (Text)
+import qualified Data.Text as Text
+import           Data.String
+import qualified Data.Vector as V
+
+import qualified Text.Regex as Regex
+import           Text.PrettyPrint.ANSI.Leijen hiding ((<$>))
+
+import           Data.Parameterized.Some
+import           Data.Parameterized.Classes
+import           Data.Parameterized.Context
+import           Data.Parameterized.TraversableFC
+import qualified Data.Parameterized.TH.GADT as U
+import qualified Data.Parameterized.Map as MapF
+import qualified Data.Parameterized.NatRepr as N
+
+import           Lang.Crucible.Backend
+import           Lang.Crucible.CFG.Expr
+import           Lang.Crucible.CFG.Extension.Safety(AssertionClassifier,NoAssertionClassifier,HasStructuredAssertions(..))
+import           Lang.Crucible.CFG.Generator hiding (dropRef)
+import           Lang.Crucible.FunctionHandle
+import           Lang.Crucible.Panic
+import           Lang.Crucible.Syntax
+import           Lang.Crucible.Types
+import           Lang.Crucible.Simulator.ExecutionTree hiding (FnState)
+import           Lang.Crucible.Simulator.Evaluation
+import           Lang.Crucible.Simulator.GlobalState
+import           Lang.Crucible.Simulator.Intrinsics
+import           Lang.Crucible.Simulator.RegValue
+import           Lang.Crucible.Simulator.RegMap
+import           Lang.Crucible.Simulator.SimError
+import           Lang.Crucible.Utils.MuxTree
+
+import           What4.Concrete (ConcreteVal(..), concreteType)
+import           What4.Interface
+import           What4.Partial
+    (PartExpr, pattern Unassigned, pattern PE, maybePartExpr, justPartExpr, joinMaybePE,
+        mergePartial, mkPE)
+import           What4.Utils.MonadST
+
+import           Mir.DefId
+import           Mir.Mir
+import           Mir.PP
+
+import           Debug.Trace
+
+import           Unsafe.Coerce
+
+
+-- | A fancier version of MuxTree, where the value type `a` can include
+-- symbolic expressions.
+--
+-- The value type must have a notion of a "concrete skeleton" (that is, it must
+-- have an `OrdSkel` instance), such that values with matching skeletons (those
+-- where `compareSkel x y == EQ`) can be muxed together.  This is effectively
+-- like splitting the values into separate concrete and symbolic parts, so we
+-- can handle the concrete parts with a MuxTree and the symbolic parts with
+-- ordinary muxing.
+--
+-- Unlike with MuxTree, it is not guaranteed that the `Pred`s of a FancyMuxTree
+-- cover all possibilities.  However, the process of constructing such a
+-- partial FancyMuxTree will assert as a side effect that the "missing" cases
+-- do not occur.  This is handled in runMuxLeafIO: if the MuxLeafT computation
+-- aborts, then runMuxLeafIO asserts that the leaf is not active before
+-- returning Nothing.
+newtype FancyMuxTree sym a = FancyMuxTree (Map (Skeleton a) (a, Pred sym))
+
+
+class OrdSkel t where
+    compareSkel :: t -> t -> Ordering
+
+class OrdSkel1 (t :: k1 -> Type) where
```

Comment:
Do `OrdSkel1`, etc, need to be classes? As far as I can tell, they're only used to eventually build an implementation of `OrdSkel` below, and could just as well be regular functions in context.

Reply:
Good point - I've removed them